modules/perception/camera/app/cipv_camera.cc

```diff
@@ -0,0 +1,967 @@
+                                         LaneLineSimple *right_lane_line) {
+  // TODO(techoe): Use union of lane line and yaw_rate path to decide the
+  // virtual lane
+  float x = 0.0f;
+  float y = 0.0f;
+  left_lane_line->line_point.clear();
+  right_lane_line->line_point.clear();
+
+  if (b_image_based_cipv_ == false) {
+    for (uint32_t i = 0; i < 120; i += 5) {
+      VehicleDynamics(i, yaw_rate, velocity, time_unit_, &x, &y);
+      Eigen::Vector2f left_point(x, y + offset_distance);
+      left_lane_line->line_point.push_back(left_point);
+      Eigen::Vector2f right_point(x, y - offset_distance);
+      right_lane_line->line_point.push_back(right_point);
+    }
+  } else {
+    // Image based extension requires to reproject virtual laneline points to
+    // image space.
+  }
+  return true;
+}
+
+// Elongate lane line
+bool Cipv::ElongateEgoLane(const std::vector<base::LaneLine> &lane_objects,
+                           const bool b_left_valid, const bool b_right_valid,
+                           const float yaw_rate, const float velocity,
+                           EgoLane *egolane_image, EgoLane *egolane_ground) {
+  float offset_distance = EGO_CAR_HALF_VIRTUAL_LANE;
+  // When left lane line is available
+  if (b_left_valid && b_right_valid) {
+    // elongate both lanes or do nothing
+    if (debug_level_ >= 2) {
+      AINFO << "Both lanes are fine";
+    }
+    // When only left lane line is available
+  } else if (!b_left_valid && b_right_valid) {
+    // Generate virtual left lane based on right lane
+    offset_distance =
+      -static_cast<float>(fabs(egolane_ground->right_line.line_point[0][1]) +
+      EGO_CAR_HALF_VIRTUAL_LANE);
+    MakeVirtualLane(egolane_ground->right_line, yaw_rate, offset_distance,
+                    &egolane_ground->left_line);
+    if (debug_level_ >= 2) {
+      AINFO << "Made left lane with offset: " << offset_distance;
+    }
+
+    // When only right lane line is available
+  } else if (b_left_valid && !b_right_valid) {
+    // Generate virtual right lane based on left lane
+    offset_distance =
+      static_cast<float>(fabs(egolane_ground->left_line.line_point[0][1]) +
+      EGO_CAR_HALF_VIRTUAL_LANE);
+    MakeVirtualLane(egolane_ground->left_line, yaw_rate, offset_distance,
+                    &egolane_ground->right_line);
+    if (debug_level_ >= 2) {
+      AINFO << "Made right lane with offset: " << offset_distance;
+    }
+
+    // When there is no lane lines available
+  } else {  // if (!b_left_valid && !b_right_valid)
+    // Generate new egolane using yaw-rate and yaw-angle
+    MakeVirtualEgoLaneFromYawRate(yaw_rate, velocity, offset_distance,
+                                  &egolane_ground->left_line,
+                                  &egolane_ground->right_line);
+    if (debug_level_ >= 2) {
+      AINFO << "Made both lane_objects";
+    }
+  }
+
+  return true;
+}
+
+// Get closest edge of an object in image coordinate
+bool Cipv::FindClosestEdgeOfObjectImage(
```

Comment:
it seems this function is used only for objects cpiv determination in image space. please add some comment for todo item

Reply:
also, pls add unit test for this function

Reply:
I added two method image-based and ground-based. Currently I am using ground-based. However, I plan to use image-based one when sensors are not calibrated. I will add unit test in next PR.